src/time.rs

```diff
@@ -0,0 +1,136 @@
+use core::{ops, fmt};
+
+static MILLIS_PER_SEC: u64 = 1_000;
+
+/// A representation of an absolute time value.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Instant {
+    millis: u64,
+}
+
+impl Instant {
+    /// Create a new `Instant`.
+    pub fn new(millis: u64) -> Instant {
+        Instant { millis }
+    }
+
+    /// Add a [Duration] to the current timestamp value and return a
+    /// `Option` containing a new [Instant] of the result on success.
+    ///
+    /// [Duration]: struct.Duration.html
+    /// [Instant]: struct.Instant.html
+    pub fn checked_add(&self, rhs: Duration) -> Option<Instant> {
+        self.millis.checked_add(rhs.millis()).map(|x| Instant::new(x))
+    }
+
+    /// Subtract a [Duration] from the current timestamp value and
+    /// return a `Option` containing a new [Instant] of the result on
+    /// success.
+    ///
+    /// [Duration]: struct.Duration.html
+    /// [Instant]: struct.Instant.html
+    pub fn checked_sub(&self, rhs: Duration) -> Option<Instant> {
+        self.millis.checked_sub(rhs.millis()).map(|x| Instant::new(x))
+    }
+
+    /// Perform a saturating subtraction with a [Duration] from
+    /// the current timestamp value and return a new [Instant] of
+    /// the result.
+    ///
+    /// [Duration]: struct.Duration.html
+    /// [Instant]: struct.Instant.html
+    pub fn saturating_sub(&self, interval: Duration) -> Instant {
+        Instant::new(self.millis.saturating_sub(interval.millis()))
+    }
+
+
+    /// The number of milliseconds that have passed since an arbitrary
+    /// point in time.
+    pub fn millis(&self) -> u64 {
+        self.millis
+    }
+
+    /// The number of seconds that have passed since an arbitrary
+    /// point in time.
+    pub fn secs(&self) -> u64 {
+        self.millis % MILLIS_PER_SEC
+    }
+}
+
+impl fmt::Display for Instant {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}s{}ms", self.secs(), self.millis())
+    }
+}
+
+impl ops::Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, rhs: Duration) -> Instant {
+        Instant::new(self.millis + rhs.millis())
+    }
+}
+
+impl ops::Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, rhs: Duration) -> Instant {
+        Instant::new(self.millis - rhs.millis())
+    }
+}
+
+/// A relative amount of time.
+#[derive(Debug, Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Duration {
+    millis: u64,
+}
+
+impl Duration {
+    /// Create a new `Duration`.
+    pub fn new(millis: u64) -> Duration {
+        Duration { millis }
+    }
+
+    /// The relative number of seconds that have passed since an
+    /// arbitrary point in time.
+    pub fn secs(&self) -> u64 {
+        self.millis
+    }
+
+    /// The relative number of milliseconds that have passed since an
+    /// arbitrary point in time.
+    pub fn millis(&self) -> u64 {
+        self.millis % MILLIS_PER_SEC
```

Comment:
It seems odd that Instant and Duration behave differently with regard to millis() and secs(). I think this should just return millis or Instant should be changed to return the modulus.